feat: Editorial review: Attribution reporting API (#33172)
mdn/content: files/en-us/web/http/headers/attribution-reporting-register-source/index.md

    </tr>
  </tbody>
</table>

## Syntax

```http
Attribution-Reporting-Register-Source: <json-string>
```

## Directives

- `<json-string>`

  - : A JSON string providing the information that the browser should store when the attribution source is interacted with. Available fields are as follows:

    - `"source_event_id"` {{optional_inline}}
      - : A string representing an ID for the attribution source, which can be used to map it to other information when the attribution source is interacted with, or aggregate information at the reporting endpoint. The string must consist solely of a base-10-formatted 64-bit unsigned integer.
    - `"destination"`
      - : A single string or an array of 1–3 strings. These strings must contain a complete URL corresponding to the site (scheme + [eTLD+1](/en-US/docs/Glossary/eTLD)) on which a trigger is expected to occur. These are used to match the attribution trigger to the source when a trigger is interacted with.
    - `"aggregation_keys"` {{optional_inline}}
      - : An object containing user-provided keys representing different data points to aggregate report values under.
    - `"aggregatable_report_window"` {{optional_inline}}

>       - : A string representing a time in seconds after which trigger data will no longer be included in generated aggregatable reports. The reports are still generated at the same time. If not set, this defaults to the `"expiry"` value.

Reviewer: What does "The reports are still generated at the same time." mean? I.e. perhaps a report is generated for the data, but not the agggreatable report, or perhaps the report will not include this data?

Author: Bleh, that doesn't make much sense, and I don't think it needs to be there. Removed.

    - `"event_report_window"` {{optional_inline}}
      - : A string representing a time in seconds, after which subsequent triggers won't be attributable to this source for the purpose of producing event-level reports. This is called a **report window**. If not set, the event report window falls back to the `"expiry"` value.
        > **Note:** If `"event_report_window"` is specified, `"event_report_windows"` cannot be specified, otherwise the source registration will fail.
    - `"event_report_windows"` {{optional_inline}}
      - : An object representing a series of report windows, starting at `"start_time"`, with reports for this source being delivered after each specified end time in `"end_times"`. This can be used to vary the time of report delivery across multiple reports. If not set, the event report window falls back to the `"expiry"` value. Properties are as follows:

>         - `"start_time"` {{optional_inline}}: A non-negative number specifying the start time for the resporting windows. If not specified, it defaults to `0`.

Reviewer 2: ```suggestion
        - `"start_time"` {{optional_inline}}: A non-negative number specifying the start time for the reporting windows. If not specified, it defaults to `0`.
```

        - `"end_times"`: An array of positive numbers specifying end times for subsequent report windows. The values must be increasing, and greater than `"start_time"`.

>           > **Note:** If `"event_report_windows"` is specified, `"event_report_window"` cannot be specified, otherwise the source registration will fail.

Reviewer: Indent wrong
```suggestion

        > **Note:** If `"event_report_windows"` is specified, `"event_report_window"` cannot be specified, otherwise the source registration will fail.
```

Author: Ah yes, fixed - thanks!

    - `"trigger_data"` {{optional_inline}}
      - : An array of 32-bit unsigned integers representing data that must be matched against `"trigger_data"` specified in [triggers](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#trigger_data) for event-level attribution to take place. If omitted, `"trigger_data"` defaults to `[0, 1, 2, 3, 4, 5, 6, 7]` for navigation-based sources and `[0, 1]` for event-based (image- or script-based) sources.
    - `"trigger_data_matching"` {{optional_inline}}

      - : A string that specifies how the `"trigger_data"` from the trigger is matched against the source's `"trigger_data"`. Possible values are:

        - `"exact"`: The `"trigger_data"` from the trigger must exactly match a value contained in the source's `"trigger_data"`; if there is no such match, no event-level attribution takes place.

>         - `"modulus"`: The source's `"trigger_data"` must form a contiguous sequence of integers starting at 0. The trigger's `"trigger_data"` is taken modulus the cardinality of this sequence and then matched against the source `"trigger_data"`. It is an error to use `"modulus"` if the trigger data does not form such a sequence.

Reviewer: I can't parse this yet. Maybe later when I've read the rest. But it reads as WTF.

Author: I'll be honest; I don't know what this means either. I copied it from the [explainer](https://github.com/WICG/attribution-reporting-api/blob/main/EVENT.md).

@apasel422, could you explain this in simpler words, show a quick example of what this might look like, and give us an idea of when the two approaches might be used?

Author: I've updated this description now; I think it is a bit more understandable now.

    - `"trigger_data"` {{optional_inline}}

      - : An array of 32-bit unsigned integers representing data that describes the different trigger events that could match this source. For example, "user added item to shopping cart" or "user signed up to mailing list" could be actions happening at the trigger site that could match this source and indicate a conversion of some kind that the advertiser is trying to measure. These must be matched against `"trigger_data"` specified in [triggers](/en-US/docs/Web/HTTP/Headers/Attribution-Reporting-Register-Trigger#trigger_data) for event-level attribution to take place. If omitted, `"trigger_data"` defaults to `[0, 1, 2, 3, 4, 5, 6, 7]` for navigation-based sources and `[0, 1]` for event-based (image- or script-based) sources.

>         > **Note:** The numbers that are used to represent each event are arbitrary, and up to you as the developer. They will be matched to the specific events you are trying to measure by the reporting server.

Reviewer: I am not sure it is the reporting server. Isn't it the browser when matching attribution triggers?

```suggestion
        > **Note:** The numbers that are used to represent each event, and the number of events in the array, are completely arbitrary, and defined by you as the developer. The array may contain events that are not used, but events must be present here in order to be attributed to the source when a trigger is registered.
```

Author: Duh, yes, you are right, it IS the browser. I mostly like your wording, but I've tweaked it a bit, and used a few words that I thought were more appropriate. In particular, I've used "values" rather than "numbers", as the values can contain other characters too, e.g. "event1".

I ended up changing the notes to:

> **Note:** The values used to represent each event, and the number of elements in the array, are completely arbitrary and defined by you as the developer. The array may contain values that are not used, but values must be present in the array to be attributed to the source by the browser when a trigger is registered.

Reviewer: Heaps better!

      - : A string that specifies how the `"trigger_data"` from the trigger is matched against the source's `"trigger_data"`. Possible values are:

        - `"exact"`: The `"trigger_data"` from the trigger must exactly match a value contained in the source's `"trigger_data"`; if there is no such match, no event-level attribution takes place.
        - `"modulus"`: In this case, the following calculation is performed — `d % allowedValues.size` — where `d` is the `"trigger_data"` from the trigger, and `allowedValues` is the sequence of values in the source's `"trigger_data"` array. If the result of this calculation matches a value in the source's `"trigger_data"` array, the match is a success. In such a case, the value will always match, unless `allowedValues` is empty.

>         `"modulus"` mode exists primarily for backwards compatibility with the API's behavior before `"exact"` was introduced, and as such, you'd be unlikely to use it. It is still useful in particular cases that require a very specific kind of compression resulting in smaller registration headers. This can be required when using complex filtering logic that needs to set different trigger data based on the source type according to the maximum number of source `"trigger_data"` items.

Reviewer: This implies that our general advice when we first introduce this is that `"trigger_data_matching` should be set to `exact`, and if we mention modulus we should  say something like "use of modulus is required in specialist cases" with a link to this explanation. Something to check - because on previous reads it wasn't clear which you should normally use

Author: This is a good point; I've gone through and checked. There is only really one other page that mentions it, the reporting page, and I've updated the code comment there to

> // Could also be "modulus", but "exact" will be used in most cases

Reviewer: Cool. I'll have a look through - but if all examples show exact being used then we're gold.

        > **Note:** If `"modulus"` is used, the source's `"trigger_data"` must form a contiguous sequence of integers starting at 0. If the trigger data does not form such a sequence, an error occurs.

        If not specified, `"trigger_data_matching"` defaults to `"modulus"`. Again, the reason for this is backwards compatibility: omitting the `"trigger_data_matching"` field needs to result in the same behavior observed before this field was introduced.

> ## Examples

Reviewer: Examples should include trigger_data_matching set to exact by default
